server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/MaxRetryAllocationDecider.java

```diff
@@ -41,27 +41,36 @@
 
     public static final String NAME = "max_retry";
 
+    private static final Decision YES_NO_FAILURES = Decision.single(Decision.Type.YES, NAME, "shard has no previous failures");
+
     @Override
     public Decision canAllocate(ShardRouting shardRouting, RoutingAllocation allocation) {
         final UnassignedInfo unassignedInfo = shardRouting.unassignedInfo();
         final Decision decision;
-        if (unassignedInfo != null && unassignedInfo.getNumFailedAllocations() > 0) {
+        final boolean debug = allocation.debugDecision();
+        final int numFailedAllocations = unassignedInfo == null ? 0 : unassignedInfo.getNumFailedAllocations();
+        if (numFailedAllocations > 0) {
             final IndexMetadata indexMetadata = allocation.metadata().getIndexSafe(shardRouting.index());
```

Comment:
I wonder if it was just as good (or better) to just extract the non-happy path here out into a method of its own?

Reply:
Sure why not, certainly fits in with the theme of this PR :)